### DISCUSSION TITLE
Setup Help for 1 Head and 7 Tool Heads

### DISCUSSION THREAD

**Discussion by Rickstar27** [TypQxQ/KTC#41] [General]
I just wanted to say, fantastic job on this plugin! You've done an excellent job, and it looks really well-made.

I have a question regarding its compatibility with my setup as I transition from Marlin to Klipper. Initially, I planned to create my own plugin but came across yours, which looks very promising.

In my setup, the main head on the carriage uses pogo pins to connect to the toolhead and locks them into place with a servo. This is currently configured in Klipper as an extruder.

Each toolhead also has pogo pins on both sides: one side connects to the dock, and the other connects to the main head. The docks are set up in Klipper as heater_generic and have their own temperature, heating, and fan pins, which can be individually controlled for each toolhead.

Here’s what I’m looking for: I’d like the tools to remain heated while docked. When the main head goes to pick up a toolhead, it should set the toolhead and main head temperature to 0 (turning it off). After dropping off the toolhead, the system should preheat the dropped-off toolhead. Meanwhile, the temperature for the newly picked-up toolhead should turn off before collecting and once collected allowing the main head to resume at the set temperature.

Would your plugin support this functionality?

**Comment by TypQxQ** [comment]
I believe it can handle that by different methods but I see some problems. At the moment of docking it would connect both pins of dock and toolhead. I don't know how the MCUs respond to that. Also when disconnecting from dock you would loose the thermistor causing errors.
If this would be electrical feasible I would go with RRF instead of Klipper because it can dynamically reallocate pins at runtime.
Klipper couldn't reallocate to other pins or disable pins at runtime last I checked.

**Comment by Rickstar27** [comment]
I've already had the issue of the thermistor being disconnected and throwing ADC error i've managed to fix that so that is not a problem now, and FYI i've created my own controller PCB that handles all the dock heads running klipper firmware.  as long as i don't send the voltage down the wrong pin i should be fine.

**Comment by TypQxQ** [comment]
That is really cool that you designed the PCBs!
I've been giving this some thought and checked the code a bit for best implementation.
You should be able to do this by simply specifying the toolhead heater name as well as the dock heater name in the `heater:` parameter of the respective tool:
```
[ktc_tool 0]
tool_number: 0
heater: extruder, extruder55
fans: partfan_t0
params_park:198,50, 0

[ktc_tool 1]
tool_number: 1
heater: extruder1, extruder55
fans: partfan_t1
params_park:198,150, 0
```
Here  extruder55 being the extruder for the toolhead.

This will set the temperatures for both the dock and toolhead heater.

---
You mentioned that you wanted to create your own plugin so I asume you have som programming knowhow so here is my logic:
- Each KtcTool object can have one KtcToolExtruder object.
- Each KtcToolExtruder has a list of KtcHeaterSettings, each having the name of a KtcHeater object and a temperature_offset for this KtcToolExtruder. This is so a Heater can be used on multiple tools.
- Each KtcHeater has a variable self.klippy_heater that corresponds to a heater defined by Klipper, for example extruder1.
- Whan setting temperature or state for a KtcTool, the code will do it for all KtcHeaters associated with the KtcToolExtruder that the KtcTool has taking in account if the tools state is active or standby.

---
One problem could be depending on the temperature reported by the dock heater when waiting for temperatures as it waits for all heaters connected to that extruder.

Am I making sense? Getting a bit late 😄

**Reply by Rickstar27** [reply]
Thanks for the fast response and the help.

So extruder55 will be the head and then all other tools will be extruder1, extruder2, ect...?

Yeah I think I get what you are saying about temperature reported, So as it takes a head from the dock because its not there anymore it cannot report the temperature / reporting the wrong temperature so then it basically be waiting and till that dock heats up that it will never do is that correct?

**Reply by TypQxQ** [reply]
It will wait forever if not giving a temperature out of range error first.
You might be able to rewrite the Klipper code to change pins at runtime.
